dev-python/QtPy/QtPy-1.9.0-r3.ebuild

```diff
@@ -0,0 +1,84 @@
+# Copyright 1999-2021 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=7
+
+PYTHON_COMPAT=( python3_{7..8} )
+
+inherit distutils-r1 virtualx
+
+DESCRIPTION="Abstraction layer on top of PyQt5 and PySide2 and additional custom QWidgets"
+HOMEPAGE="https://github.com/spyder-ide/qtpy"
+SRC_URI="mirror://pypi/${PN:0:1}/${PN}/${P}.tar.gz"
+
+LICENSE="MIT"
+SLOT="0"
+KEYWORDS="~amd64 ~arm64 ~x86"
+
+IUSE="
+	3d bluetooth charts concurrent datavis dbus declarative designer
+	examples gles2-only gui help location multimedia network networkauth
+	opengl positioning printsupport qml quick script scripttools
+	scxml sensors serialport speech sql ssl svg test testlib webchannel
+	webengine webkit websockets widgets x11extras xml xmlpatterns
```

Comment:
I honestyl don't see the point of all these USE flags.

Reply:
I understand that it is not desirable to have a ton of USE flags. However, I don't see an alternative here, say I have a package foo which requires Qt Python bindings for dev-qt/gui, dev-qt/printsupport, and dev-qt/qtxml, in the current situation I could just:
```
DEPEND="dev-python/QtPy[gui,printsupport,xml]"
```
and all will be taken care of. If I were to reduce the number of flags and e.g. remove the `printsupport` and xml flag. I would have to:
```
DEPEND="dev-python/QtPy[gui]
    || (
        dev-python/PyQt5[printsupport]
        dev-python/pyside[printsupport,xml]
   )"
```
And things get more complicated, the more qt modules a Python package needs.

What I have been trying to do here is to have this ebuild act as a abstraction layer for PySide2 and PyQt5, so any ebuild can just depend on QtPy with the correct flags and then the ebuild will take care of pulling in the correct packages providing the necessary Python bindings. Otherwise it would be necessary for every ebuild maintainer to figure out which flags provide which Python bindings and which things are installed unconditionally, and to create a `||` statement in every ebuild depending on QtPy. 

In essence, this ebuild acts as a virtual for PyQt5/PySide2 (though it technically is not because it does install some files). And I don't think the large amount of flags is a problem in this case because the compile time of the package is trivially low.

I'm open to other suggestions, but at the moment I am convinced that this is the best way to do it because this way it is not necessary for every ebuild depending on QtPy to have a `||` statement pulling in the correct configuration of PySide2 or PyQt5.

Reply:
@juippis just a thought I had this morning, what if I'd add an actual virtual for QtPy and move the dependency stuff and USE flags in there. So `virtual/QtPy` would have:
```
RDEPEND="
	app-eselect/eselect-QtPy
        dev-python/QtPy
	|| (
		dev-python/PyQt5[${PYTHON_USEDEP},designer?,gles2-only?,gui?,help?,location?,multimedia?,network?,opengl?,positioning?,printsupport?,sensors?,serialport?,sql?,svg?,testlib?,webchannel?,websockets?,widgets?,x11extras?,xml(+)?,xmlpatterns?]
		dev-python/pyside2[${PYTHON_USEDEP},designer?,gles2-only?,gui?,help?,location?,multimedia?,network?,opengl(+)?,positioning?,printsupport?,sensors?,serialport(+)?,sql?,svg?,testlib?,webchannel?,websockets?,widgets?,x11extras?,xml?,xmlpatterns?]
	)
	webengine? ( || (
		dev-python/PyQtWebEngine[${PYTHON_USEDEP}]
		dev-python/pyside2[${PYTHON_USEDEP},webengine]
	) )
	qml? ( || (
		dev-python/PyQt5[${PYTHON_USEDEP},declarative]
		dev-python/pyside2[${PYTHON_USEDEP},qml]
	) )
	quick? ( || (
		dev-python/PyQt5[${PYTHON_USEDEP},declarative]
		dev-python/pyside2[${PYTHON_USEDEP},quick]
	) )
	declarative? ( || (
		dev-python/PyQt5[${PYTHON_USEDEP},declarative]
		dev-python/pyside2[${PYTHON_USEDEP},qml,quick]
	) )
	bluetooth? ( dev-python/PyQt5[${PYTHON_USEDEP},bluetooth] )
	dbus? ( dev-python/PyQt5[${PYTHON_USEDEP},dbus] )
	examples? ( dev-python/PyQt5[${PYTHON_USEDEP},examples] )
	networkauth? ( dev-python/PyQt5[${PYTHON_USEDEP},networkauth] )
	ssl? ( dev-python/PyQt5[${PYTHON_USEDEP},ssl] )
	webkit? ( dev-python/PyQt5[${PYTHON_USEDEP},webkit] )
	3d? ( dev-python/pyside2[${PYTHON_USEDEP},3d] )
	charts? ( dev-python/pyside2[${PYTHON_USEDEP},charts] )
	concurrent? ( dev-python/pyside2[${PYTHON_USEDEP},concurrent] )
	datavis? ( dev-python/pyside2[${PYTHON_USEDEP},datavis] )
	scxml? ( dev-python/pyside2[${PYTHON_USEDEP},scxml] )
	script? ( dev-python/pyside2[${PYTHON_USEDEP},script] )
	scripttools? ( dev-python/pyside2[${PYTHON_USEDEP},scripttools] )
	speech? ( dev-python/pyside2[${PYTHON_USEDEP},speech] )
"
```
So like this `dev-python/QtPy` does not have the dependency and USE flag mess and does not have to depend on anything at all. And the USE flags would be on an actual virtual instead of directly on QtPy itself.

Would that be better in your opinion?

Reply:
I guess the `src_test` in this ebuild fooled me a bit earlier, thinking this has some runnable code, hence also clinging to the USE flag mess? 

Since the functionality is controlled via dependencies, maybe the cleanest and working solution would just to be print a message about it and leave it up to users? 

I'd leave the "engine choices" to this ebuild like they are in the current in-tree version, but if a package requires "QtPy" "dbus" support, just directly depend on `dev-python/QtPy[${PYTHON_USEDEP}]` and `dev-python/PyQt5[${PYTHON_USEDEP},dbus]`. Or simply the latter. 
I guess the only downside I see to that is that if pyside2 and PyQt5 start getting more cross-support, but how likely would that be? They seem, rather consistent.

I'm also not so sure about the eselect module, these seem to be able to co-exist. 

This may be one of those cases where there is no good solution to solve it. Yours isn't horrible either, but when looking into future, I can see it become more of a burden, and it looks risky if you know what I mean :P

Reply:
> I'd leave the "engine choices" to this ebuild like they are in the current in-tree version, but if a package requires "QtPy" "dbus" support, just directly depend on `dev-python/QtPy[${PYTHON_USEDEP}]` and `dev-python/PyQt5[${PYTHON_USEDEP},dbus]`. Or simply the latter.
> I guess the only downside I see to that is that if pyside2 and PyQt5 start getting more cross-support, but how likely would that be? They seem, rather consistent.

So keep only those things that are common to both PyQt5 and PySide2? Sure, we could do that.
if more cross-support is added in the future we can just add those flags later, I'll leave a comment in the ebuild reminding myself to periodically check that when bumping it.

> I'm also not so sure about the eselect module, these seem to be able to co-exist.

They can co-exist on the system, but a given application can only use one of them at the same time, which is controlled by the `QT_API` environment variable (which can be set with the eselect-qtpy module).

QtPy will use whatever is installed, but if both are installed it will default to PyQt5 over PySide2 (unless `QT_API` is set, which forces it to use whatever is in that variable).

It is mostly useful for cases where you have both installed on the system and a given application is buggy with one binding and you therefore want to explicitly use the other one.

Personally, I think it is very useful, rather than having to set `QT_API="pyside2"` manually in some file that will get sourced and than forgetting which file I put it.

Reply:
@juippis 
I just updated the ebuild and dropped all the flags that were not provided by both PyQt5 and PySide2, so now it is more like the current in-tree ebuild but expanded. I also added some comments in the ebuild explaining why things are the way they are, which should hopefully make it easier to maintain in the future.

Let me know what you think :)